File: C++/merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
	//Fucntion to append node at the end of lisked list
    struct ListNode *begg(struct ListNode *head, int x){
        struct ListNode *temp = new ListNode(x);
        //If there is no node
        if(head==NULL){
            head=temp;
            return head;
        }
        //If nodes already present
        struct ListNode *ptr=head;
        while(ptr->next){
            ptr=ptr->next;
        }
        ptr->next=temp;
        return head;
    }
    ListNode* mergeKLists(vector<ListNode*>& lists) {
    	//Using multiset s it will allow duplicate values and will store them in sorted order.
        multiset<int> s;
        //Adding values to multiset.
        for(int i=0;i<lists.size();i++){
            ListNode *p1=lists[i];
            while(p1){
                s.insert(p1->val);
                p1=p1->next;
            }
        }
        //Initializing the head of linked list.
        struct ListNode *head2=NULL;
        //Appending nodes in the linked list from multiset.
        for(auto it=s.begin();it!=s.end();++it){
            head2=begg(head2,*it);
            cout<<*it<<" ";
        }
        return head2;
    }
};

File: Python/merge-k-sorted-lists.py
# https://leetcode.com/problems/merge-k-sorted-lists/
# Difficulty Level: Hard
# Tags: Math
# Time Complexity: O(log(n))
# Space Complexity: O(n)


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        
        if lists == []:
            return
        
        if len(lists) == 0:
            return []
        
        size = len(lists)
        
        def merge2(list1, list2):
            
            x = ListNode(0)
            y = x
            
            while (list1 and list2):
                
                if list1:
                    
                    if list1.val > list2.val:
                        
                        y.next = list2
                        list2 = list2.next
                        
                    else:
                        
                        y.next = list1
                        list1 = list1.next
                        
                    y = y.next
                    
                else:
                    
                    if list1.val < list2.val:
                        
                        y.next = list1
                        list2 = list1.next
                        
                    else:
                        
                        y.next = list2
                        list2 = list2.next
                        
                    y = y.next
                    
            if list1:
                
                y.next = list1
                
            elif list2:
                
                y.next = list2
                
            return x.next
        
        index = 1
        merged = lists[0]
        
        while (index < size):
            
            merged = merge2(merged, lists[index])
            index = index + 1
            
        return merged